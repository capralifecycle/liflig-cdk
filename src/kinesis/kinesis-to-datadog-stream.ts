import * as cdk from "aws-cdk-lib"
import * as iam from "aws-cdk-lib/aws-iam"
import * as firehose from "aws-cdk-lib/aws-kinesisfirehose"
import * as logs from "aws-cdk-lib/aws-logs"
import * as s3 from "aws-cdk-lib/aws-s3"
import { BlockPublicAccess } from "aws-cdk-lib/aws-s3"
import * as secretsmanager from "aws-cdk-lib/aws-secretsmanager"
import * as constructs from "constructs"

export interface KinesisToDatadogStreamProps {
  /**
   *
   * The name of the SecretsManager secret where your Datadog API key is saved.
   *
   * The secret must be a JSON object on the format { "value": "SECRET" }
   *
   */
  datadogApiKeySecretName: string
  /**
   *
   * The CloudWatch log groups from you are streaming to Datadog
   *
   */
  logGroups: logs.ILogGroup[]
}

/**
 *
 * Forwards logs from log-groups in CloudWatch to a Datadog account.
 * The logs are delivered through a Firehose delivery stream, which is being subscribed to the log-groups in CloudWatch.
 *
 * @author Stein-Aage
 */
export class KinesisToDatadogStream extends constructs.Construct {
  constructor(
    scope: constructs.Construct,
    id: string,
    props: KinesisToDatadogStreamProps,
  ) {
    super(scope, id)

    const deliveryStreamLogGroup = new logs.LogGroup(
      this,
      "DeliveryStreamLogGroup",
    )

    const deliveryStreamLogStream = new logs.LogStream(
      this,
      "DeliveryStreamLogStream",
      {
        logGroup: deliveryStreamLogGroup,
      },
    )

    const failedDataBucket = new s3.Bucket(this, "FailedDataBucket", {
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
    })

    const cloudWatchLogsRole = new iam.Role(this, "CloudWatchLogsRole", {
      assumedBy: new iam.ServicePrincipal(
        `logs.${cdk.Stack.of(this).region}.amazonaws.com`,
      ),
    })

    const firehoseLogsRole = new iam.Role(this, "FirehoseLogsRole", {
      assumedBy: new iam.ServicePrincipal("firehose.amazonaws.com"),
    })

    const datadogDeliveryStream = new firehose.CfnDeliveryStream(
      this,
      "DeliveryStream",
      {
        deliveryStreamType: "DirectPut",
        httpEndpointDestinationConfiguration: {
          roleArn: firehoseLogsRole.roleArn,
          endpointConfiguration: {
            url: "https://aws-kinesis-http-intake.logs.datadoghq.eu/v1/input",
            accessKey: secretsmanager.Secret.fromSecretNameV2(
              scope,
              "DatadogApiKey",
              props.datadogApiKeySecretName,
            )
              .secretValueFromJson("value")
              .toString(),
            name: "datadog-logs-endpoint",
          },
          requestConfiguration: {
            contentEncoding: "GZIP",
          },
          cloudWatchLoggingOptions: {
            enabled: true,
            logGroupName: deliveryStreamLogGroup.logGroupName,
            logStreamName: deliveryStreamLogStream.logStreamName,
          },
          bufferingHints: {
            intervalInSeconds: 60,
            sizeInMBs: 4,
          },
          retryOptions: {
            durationInSeconds: 60,
          },
          s3BackupMode: "FailedDataOnly",
          s3Configuration: {
            bucketArn: failedDataBucket.bucketArn,
            compressionFormat: "UNCOMPRESSED",
            roleArn: firehoseLogsRole.roleArn,
          },
        },
      },
    )

    new iam.Policy(this, "CloudWatchLogsPolicy", {
      document: new iam.PolicyDocument({
        statements: [
          new iam.PolicyStatement({
            actions: ["firehose:PutRecord", "firehose:PutRecordBatch"],
            resources: [datadogDeliveryStream.attrArn],
          }),
        ],
      }),
      roles: [cloudWatchLogsRole],
    })

    new iam.Policy(this, "FirehoseLogsPolicy", {
      document: new iam.PolicyDocument({
        statements: [
          new iam.PolicyStatement({
            actions: [
              "s3:AbortMultipartUpload",
              "s3:GetBucketLocation",
              "s3:GetObject",
              "s3:ListBucket",
              "s3:ListBucketMultipartUploads",
              "s3:PutObject",
            ],
            resources: [
              failedDataBucket.bucketArn,
              `${failedDataBucket.bucketArn}/*`,
            ],
          }),
          new iam.PolicyStatement({
            actions: ["logs:PutLogEvents"],
            resources: [
              `arn:aws:logs:${cdk.Stack.of(this).region}:${
                cdk.Stack.of(this).account
              }:log-group:${deliveryStreamLogGroup.logGroupName}:log-stream:${
                deliveryStreamLogStream.logStreamName
              }`,
            ],
          }),
          new iam.PolicyStatement({
            actions: [
              "kinesis:DescribeStream",
              "kinesis:GetShardIterator",
              "kinesis:GetRecords",
            ],
            resources: [datadogDeliveryStream.attrArn],
          }),
        ],
      }),
      roles: [firehoseLogsRole],
    })

    props.logGroups.forEach((logGroup, index) => {
      new logs.CfnSubscriptionFilter(this, `SubscriptionFilter${index}`, {
        logGroupName: logGroup.logGroupName,
        destinationArn: datadogDeliveryStream.attrArn,
        filterPattern: logs.FilterPattern.allEvents().logPatternString,
        roleArn: cloudWatchLogsRole.roleArn,
      })
    })
  }
}
