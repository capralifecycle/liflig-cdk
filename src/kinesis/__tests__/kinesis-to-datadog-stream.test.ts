import "@aws-cdk/assert/jest"
import { App, Stack } from "aws-cdk-lib"
import * as logs from "aws-cdk-lib/aws-logs"
import "jest-cdk-snapshot"
import { KinesisToDatadogStream } from "../kinesis-to-datadog-stream"

test("create kinesis stream", () => {
  const app = new App()
  const supportStack = new Stack(app, "SupportStack", {
    env: {
      region: "eu-west-1",
    },
  })
  const stack = new Stack(app, "Stack", {
    env: {
      region: "eu-west-1",
    },
  })

  const logGroup = new logs.LogGroup(supportStack, "LogGroup")

  new KinesisToDatadogStream(stack, "KinesisToDatadogStream", {
    logGroups: [logGroup],
    datadogApiKeySecretName: "DATADOG-SECRET",
  })

  expect(stack).toMatchCdkSnapshot()
})
