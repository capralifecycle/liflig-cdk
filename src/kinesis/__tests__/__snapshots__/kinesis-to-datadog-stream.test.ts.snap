// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create kinesis stream 1`] = `
Object {
  "Resources": Object {
    "KinesisToDatadogStreamCloudWatchLogsPolicy5D6050AF": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "firehose:PutRecord",
                "firehose:PutRecordBatch",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "KinesisToDatadogStreamDeliveryStream83902C8C",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "KinesisToDatadogStreamCloudWatchLogsPolicy5D6050AF",
        "Roles": Array [
          Object {
            "Ref": "KinesisToDatadogStreamCloudWatchLogsRole8BADDC70",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "KinesisToDatadogStreamCloudWatchLogsRole8BADDC70": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "logs.eu-west-1.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "KinesisToDatadogStreamDeliveryStream83902C8C": Object {
      "Properties": Object {
        "DeliveryStreamType": "DirectPut",
        "HttpEndpointDestinationConfiguration": Object {
          "BufferingHints": Object {
            "IntervalInSeconds": 60,
            "SizeInMBs": 4,
          },
          "CloudWatchLoggingOptions": Object {
            "Enabled": true,
            "LogGroupName": Object {
              "Ref": "KinesisToDatadogStreamDeliveryStreamLogGroup7179F142",
            },
            "LogStreamName": Object {
              "Ref": "KinesisToDatadogStreamDeliveryStreamLogStream32974BF1",
            },
          },
          "EndpointConfiguration": Object {
            "AccessKey": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "{{resolve:secretsmanager:arn:",
                  Object {
                    "Ref": "AWS::Partition",
                  },
                  ":secretsmanager:eu-west-1:",
                  Object {
                    "Ref": "AWS::AccountId",
                  },
                  ":secret:DATADOG-SECRET:SecretString:value::}}",
                ],
              ],
            },
            "Name": "datadog-logs-endpoint",
            "Url": "https://aws-kinesis-http-intake.logs.datadoghq.eu/v1/input",
          },
          "RequestConfiguration": Object {
            "ContentEncoding": "GZIP",
          },
          "RetryOptions": Object {
            "DurationInSeconds": 60,
          },
          "RoleARN": Object {
            "Fn::GetAtt": Array [
              "KinesisToDatadogStreamFirehoseLogsRoleDF723823",
              "Arn",
            ],
          },
          "S3BackupMode": "FailedDataOnly",
          "S3Configuration": Object {
            "BucketARN": Object {
              "Fn::GetAtt": Array [
                "KinesisToDatadogStreamFailedDataBucket95726055",
                "Arn",
              ],
            },
            "CompressionFormat": "UNCOMPRESSED",
            "RoleARN": Object {
              "Fn::GetAtt": Array [
                "KinesisToDatadogStreamFirehoseLogsRoleDF723823",
                "Arn",
              ],
            },
          },
        },
      },
      "Type": "AWS::KinesisFirehose::DeliveryStream",
    },
    "KinesisToDatadogStreamDeliveryStreamLogGroup7179F142": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "RetentionInDays": 731,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "KinesisToDatadogStreamDeliveryStreamLogStream32974BF1": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "LogGroupName": Object {
          "Ref": "KinesisToDatadogStreamDeliveryStreamLogGroup7179F142",
        },
      },
      "Type": "AWS::Logs::LogStream",
      "UpdateReplacePolicy": "Retain",
    },
    "KinesisToDatadogStreamFailedDataBucket95726055": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "KinesisToDatadogStreamFirehoseLogsPolicy6D41913D": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:AbortMultipartUpload",
                "s3:GetBucketLocation",
                "s3:GetObject",
                "s3:ListBucket",
                "s3:ListBucketMultipartUploads",
                "s3:PutObject",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "KinesisToDatadogStreamFailedDataBucket95726055",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "KinesisToDatadogStreamFailedDataBucket95726055",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "logs:PutLogEvents",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:logs:eu-west-1:",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":log-group:",
                    Object {
                      "Ref": "KinesisToDatadogStreamDeliveryStreamLogGroup7179F142",
                    },
                    ":log-stream:",
                    Object {
                      "Ref": "KinesisToDatadogStreamDeliveryStreamLogStream32974BF1",
                    },
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "kinesis:DescribeStream",
                "kinesis:GetShardIterator",
                "kinesis:GetRecords",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "KinesisToDatadogStreamDeliveryStream83902C8C",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "KinesisToDatadogStreamFirehoseLogsPolicy6D41913D",
        "Roles": Array [
          Object {
            "Ref": "KinesisToDatadogStreamFirehoseLogsRoleDF723823",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "KinesisToDatadogStreamFirehoseLogsRoleDF723823": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "firehose.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "KinesisToDatadogStreamSubscriptionFilter0D455E7D8": Object {
      "Properties": Object {
        "DestinationArn": Object {
          "Fn::GetAtt": Array [
            "KinesisToDatadogStreamDeliveryStream83902C8C",
            "Arn",
          ],
        },
        "FilterPattern": "",
        "LogGroupName": Object {
          "Fn::ImportValue": "SupportStack:ExportsOutputRefLogGroupF5B4693119CE9848",
        },
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "KinesisToDatadogStreamCloudWatchLogsRole8BADDC70",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Logs::SubscriptionFilter",
    },
  },
}
`;
